File: 238.product-of-array-except-self.js
/*
 * @lc app=leetcode id=238 lang=javascript
 *
 * [238] Product of Array Except Self
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function (nums) {
  const leftProd = Array(nums.length).fill(1);
  for (let i = 1; i < nums.length; ++i) {
    leftProd[i] = leftProd[i - 1] * nums[i - 1];
  }
  let rightProd = 1;
  for (let i = nums.length - 2; i > -1; --i) {
    rightProd *= nums[i + 1];
    leftProd[i] *= rightProd;
  }
  return leftProd;
};
// @lc code=end

File: 238.product-of-array-except-self.py
#
# @lc app=leetcode id=238 lang=python3
#
# [238] Product of Array Except Self
#

# Time:  O(n)
# Space: O(1)

from typing import List


# @lc code=start
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        left_product = [1] * len(nums)
        for i in range(1, len(nums)):
            left_product[i] = left_product[i - 1] * nums[i - 1]

        right_product = 1
        for i in range(len(nums) - 2, -1, -1):
            right_product *= nums[i + 1]
            left_product[i] *= right_product

        return left_product


# @lc code=end
